compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -1129,9 +1129,20 @@ fn should_encode_trait_impl_trait_tys(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
     })
 }
 
-// Return `false` to avoid encoding impl trait in trait, while we don't use the query.
-fn should_encode_fn_impl_trait_in_trait<'tcx>(_tcx: TyCtxt<'tcx>, _def_id: DefId) -> bool {
-    false
+// We only encode impl trait in trait when using `lower-impl-trait-in-trait-to-assoc-ty` unstable
+// option.
+fn should_encode_fn_impl_trait_in_trait<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> bool {
+    if tcx.sess.opts.unstable_opts.lower_impl_trait_in_trait_to_assoc_ty {
+        match tcx.def_kind(def_id) {
+            DefKind::Fn | DefKind::AssocFn => {
+                tcx.def_kind(tcx.parent(def_id)) == DefKind::Trait
+                    && tcx.hir().get_fn_output(def_id.expect_local()).is_some()
```

Comment:
Can you just call `opt_associated_item`, check the `container` and the `kind` there? I don't even think we need to check if the output of the fn is non-empty.